### ISSUE TITLE
Typechecker

### ISSUE THREAD

**Issue by goodmind** [yever/rusty-typescript#34]
Do you plan to convert typechecker?

**Comment by lulu-berlin** [comment]
Yes, the entire source code of TypeScript.

**Comment by goodmind** [comment]
Would you also refactor it to traits, because it is all closures in closures which is already slow in js
https://github.com/Microsoft/TypeScript/issues/17861
Splitting it into more files would be nice

**Comment by lulu-berlin** [comment]
Yes, I would like to do that. But for now I'm focusing on getting things done. This would be a really nice way to refactor the code.

As always, pull requests are super welcome :)

**Comment by Himujjal** [comment]
So. This is a complete rewrite of the TypeScript compiler which is cool.

But what is the current codebase doing? You seem to embed the TypeScript compiler. How are you executing it? QuickJS?

And is the parser complete? If the parser isnt complete we can use [tree-sitter-typescript](https://github.com/tree-sitter/tree-sitter-typescript) to get an AST. Its a done project and can perfectly create the AST as of now.
